File: common/src/main/java/org/tdf/common/serialize/Codecs.java
package org.tdf.common.serialize;

import org.tdf.rlp.RLPCodec;

import java.nio.charset.StandardCharsets;
import java.util.function.Function;

public class Codecs {
    /**
     * Converter from string to byte array and vice versa
     */
    public static final Codec<String, byte[]> STRING = Codec
            .newInstance(
                    (x) -> x.getBytes(StandardCharsets.UTF_8),
                    x -> new String(x, StandardCharsets.UTF_8)
            );
    static Codec IDENTITY = Codec.newInstance(Function.identity(), Function.identity());

    public static <K> Codec<K, byte[]> newRLPCodec(Class<K> clazz) {
        return Codec.newInstance(RLPCodec::encode, x -> RLPCodec.decode(x, clazz));
    }
}

File: common/src/main/java/org/tdf/common/serialize/Codecs.kt
package org.tdf.common.serialize

import org.tdf.common.serialize.Codec.Companion.create
import org.tdf.common.util.decode
import org.tdf.common.util.hex
import org.tdf.common.util.rlp
import java.nio.charset.StandardCharsets
import java.util.function.Function

object Codecs {
    /**
     * Converter from string to byte array and vice versa
     */
    @JvmField
    val string = create(
        { it.toByteArray(StandardCharsets.UTF_8) }
    ) { String(it, StandardCharsets.UTF_8) }

    val identity: Codec<*> = create(Function.identity(), Function.identity())

    val hex = create({ it.bytes }) { it.hex() }

    fun <K> rlp(clazz: Class<K>): Codec<K> {
        return create({ it.rlp() }) { it.decode(clazz) }
    }
}